spec/cloud.go

```diff
@@ -58,17 +61,42 @@ func SuggestCloudGenerator(conf *config.Config) *CloudGenerator {
 		return &CloudGenerator{&AzureVMGenerator{azureVMBaseURL}}
 	}
 
-	if isEC2() {
-		return &CloudGenerator{&EC2Generator{ec2BaseURL}}
-	}
-	if isGCE() {
-		return &CloudGenerator{&GCEGenerator{gceMetaURL}}
-	}
-	if isAzure() {
-		return &CloudGenerator{&AzureVMGenerator{azureVMBaseURL}}
+	var wg sync.WaitGroup
+	gCh := make(chan *CloudGenerator)
+
+	// cancelable context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg.Add(3)
+	go func() {
+		if isEC2(ctx) {
+			gCh <- &CloudGenerator{&EC2Generator{ec2BaseURL}}
+		}
+		wg.Done()
+	}()
+	go func() {
+		if isGCE(ctx) {
+			gCh <- &CloudGenerator{&GCEGenerator{gceMetaURL}}
+		}
+		wg.Done()
+	}()
+	go func() {
+		if isAzure(ctx) {
+			gCh <- &CloudGenerator{&AzureVMGenerator{azureVMBaseURL}}
+		}
+		wg.Done()
+	}()
+	go func() {
+		wg.Wait()
```

Comment:
I think wait group here does not work as expected, the first resolved CloudGenerator is returned before all `wg.Done()` is called.

Reply:
I think we don't have to wait all goroutines finish, because we only need first (and in most cases only) resolved CloudGenerator. So In original p-r I didn't wait `wg` synchronously.

But for simplification and with `retry.WithContext` (thank you!), I changed to wait synchronously as you suggest.

Reply:
☝️ I tried in https://github.com/mackerelio/mackerel-agent/pull/516/commits/25dda67882e1a6eded26c063fa699a0eab15a38e but it caused data race, so I reverted back to asynchronous `wg.Wait()` at https://github.com/mackerelio/mackerel-agent/pull/516/commits/a6e43ab12663213cbe88f3063857a34d352e327f